File: code/java/lists/src/main/java/org/swufe/datastructure/LinkedList.java
package org.swufe.datastructure;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;

/**
 * A singly linked list.
 */
public class LinkedList<Item> implements Iterable<Item> {
    private static class Node<Item> {
        Item item;
        Node<Item> next;
        Node(Item item) {
            this.item = item;
            next = null;
        }
    }

    private Node<Item> head;
    private Node<Item> tail;
    private int size;

    public LinkedList() {
        head = null;
        tail = null;
        size = 0;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public int size() {
        return size;
    }

    public void addFirst(Item item) {
        Node<Item> node = new Node<>(item);
        node.next = head;
        head = node;
        if (tail == null) {
            tail = node;
        }
        size += 1;
    }

    public void addLast(Item item) {
        Node<Item> oldTail = tail;
        tail = new Node<>(item);
        if (oldTail == null) {
            head = tail;
        } else {
            oldTail.next = tail;
        }
        size += 1;
    }

    public void removeFirst() {
        if (isEmpty()) {
            throw new NoSuchElementException();
        } else {
            head = head.next;
            if (head == null) {
                tail = null;
            }
            size -= 1;
        }
    }

    public void removeLast() {
        if (isEmpty()) {
            throw new NoSuchElementException();
        } else {
            if (head == tail) {
                // only one element
                head = null;
                tail = null;
            } else {
                // head -> ... -> walk -> tail
                Node<Item> walk = head;
                while (walk.next != tail) {
                    walk = walk.next;
                }
                walk.next = null;
                tail = walk;
            }
            size -= 1;
        }
    }

    public int indexOf(Item item) {
        int index = -1;
        Node<Item> walk = head;
        while (walk != null) {
            index += 1;
            if (walk.item.equals(item)) {
                return index;
            } else {
                walk = walk.next;
            }
        }
        return -1;
    }

    /**
     * Insert item at index i
     * @param i
     * @param item
     */
    public void insert(int i, Item item) {
        if (i < 0 || i > size) {
            throw new IndexOutOfBoundsException();
        }
        if (i == 0) {
            addFirst(item);
        } else if (i == size) {
            addLast(item);
        } else {
            // before: head -> ... -> (i-1-th node: walk) -> (i-th node) -> ... -> tail
            // after: head -> ... -> (i-1-th node: walk) -> Node(item) -> (i-th node) -> ... -> tail
            Node<Item> walk = getNode(i - 1);
            Node<Item> ith = walk.next;
            Node<Item> node = new Node<>(item);
            walk.next = node;
            node.next = ith;
            size += 1;
        }
    }

    private Node<Item> getNode(int i) {
        if (i < 0 || i >= size) {
            throw new IndexOutOfBoundsException();
        }

        int cnt = 0;
        Node<Item> walk = head;
        while (cnt < i) {
            walk = walk.next;
            cnt += 1;
        }

        return walk;
    }
    public Item get(int i) {
        Node<Item> walk = getNode(i);
        assert walk != null;
        return walk.item;
    }

    public void set(int i, Item item) {
        Node<Item> walk = getNode(i);
        assert walk != null;
        walk.item = item;
    }

    public void clear() {
        while (!isEmpty()) {
            removeFirst();
        }
    }

    /**
     * A utility method for debugging
     */
    public List<Item> toArrayList() {
        List<Item> results = new ArrayList<>();
        Node<Item> walk = head;
        while (walk != null) {
            results.add(walk.item);
            walk = walk.next;
        }
        return results;
    }

    @Override
    public Iterator<Item> iterator() {
        return new LinkedListIterator();
    }

    private class LinkedListIterator implements Iterator<Item> {
        private Node<Item> walk = head;
        @Override
        public boolean hasNext() {
            return walk != null;
        }

        @Override
        public Item next() {
            Item answer = walk.item;
            walk = walk.next;
            return answer;
        }
    }
}

File: code/python/lists/linked_list.py
class Node:
    def __init__(self, item, nxt=None):
        self.item = item
        self.next = nxt


class LinkedList:
    """A singly linked list."""

    def __init__(self):
        self._head = None
        self._tail = None
        self._size = 0

    def size(self):
        return self._size

    def is_empty(self):
        return self.size() == 0

    def add_first(self, item):
        node = Node(item)
        node.next = self._head
        self._head = node
        if self._tail is None:
            self._tail = node
        self._size += 1

    def add_last(self, item):
        old_tail = self._tail
        self._tail = Node(item)
        if old_tail is None:
            self._head = self._tail
        else:
            old_tail.next = self._tail
        self._size += 1

    def remove_first(self):
        if self.is_empty():
            raise IndexError('Remove from empty linked list')
        else:
            self._head = self._head.next
            if self._head is None:
                self._tail = None
            self._size -= 1

    def remove_last(self):
        if self.is_empty():
            raise IndexError('Remove from empty linked list')
        else:
            if self._head is self._tail:
                # only one element
                self._head = None
                self._tail = None
            else:
                walk = self._head
                while walk.next is not self._tail:
                    walk = walk.next
                walk.next = None
                self._tail = walk
            self._size -= 1

    def index_of(self, item):
        idx = -1
        walk = self._head
        while walk is not None:
            idx += 1
            if walk.item == item:
                return idx
            else:
                walk = walk.next
        return -1

    def _get_node(self, i) -> Node:
        if i < 0 or i >= self._size:
            raise IndexError('Out of bounds')
        cnt = 0
        walk = self._head
        while cnt < i:
            walk = walk.next
            cnt += 1
        return walk

    def get(self, i):
        walk = self._get_node(i)
        assert walk is not None
        return walk.item

    def set(self, i, item):
        walk = self._get_node(i)
        assert walk is not None
        walk.item = item

    def insert(self, i, item):
        if i < 0 or i > self._size:
            raise IndexError('Out of bounds')
        if i == 0:
            self.add_first(item)
        elif i == self._size:
            self.add_last(item)
        else:
            # before: head -> ... -> (i-1-th node: walk) -> (i-th node) -> ... -> tail
            # after:  head -> ... -> (i-1-th node: walk) -> Node(item) -> (i-th node) -> ... -> tail
            walk = self._get_node(i-1)
            ith = walk.next
            node = Node(item)
            walk.next = node
            node.next = ith
            self._size += 1

    def clear(self):
        while not self.is_empty():
            self.remove_first()

    def __iter__(self):
        return LinkedListIterator(self._head)


class LinkedListIterator:
    def __init__(self, head: Node):
        self.walk = head

    def __next__(self):
        if self.walk is None:
            raise StopIteration
        answer = self.walk.item
        self.walk = self.walk.next
        return answer
